File: pyNCAA/scripts/sources/espn_scoreboard.py
# scripts/sources/espn_scoreboard.py
# ESPN scoreboard for NCAA men's basketball

import re
import requests
import math


def fetch_scoreboard(date_yyyymmdd=None):
    base = "https://site.api.espn.com/apis/site/v2/sports/basketball/mens-college-basketball/scoreboard"
    params = f"?dates={date_yyyymmdd}&groups=50&limit=365" if date_yyyymmdd else "?groups=50&limit=365"
    url = f"{base}{params}"
    res = requests.get(url, headers={"user-agent": "ncaa-picks-bot/1.0"})
    if not res.ok:
        raise Exception(f"HTTP {res.status_code} for {url}")
    return res.json()


def _norm_for_match(s):
    """Normalize a team name for fuzzy comparison."""
    s = str(s or "").lower()
    s = re.sub(r"[^a-z0-9\s]", " ", s)
    s = re.sub(r"\bstate\b", "st", s)
    s = re.sub(r"\s+", " ", s).strip()
    return s


def fetch_tournament_teams(date_yyyymmdd, stats_keys=None):
    """Fetch teams playing in the NCAA tournament (groups=100) and NIT (groups=98)
    from ESPN's scoreboard for the given date.

    Returns a dict mapping normalized name variants to the best short name.
    If stats_keys (e.g. Barttorvik team names) are provided, ESPN names are
    resolved to matching stats keys for consistency.
    """
    base = "https://site.api.espn.com/apis/site/v2/sports/basketball/mens-college-basketball/scoreboard"
    teams = {}  # norm_name -> best short name

    # Build a lookup from stats_keys for resolving ESPN names -> model names
    stats_norm = {}
    if stats_keys:
        for k in stats_keys:
            stats_norm[_norm_for_match(k)] = k

    def _best_name(team_obj):
        """Resolve an ESPN team object to the best model-compatible name."""
        location = (team_obj.get("location") or "").strip()   # e.g. "Ohio State"
        display = (team_obj.get("displayName") or "").strip()  # e.g. "Ohio State Buckeyes"
        abbr = (team_obj.get("abbreviation") or "").strip()    # e.g. "OSU"

        # Try matching location against stats keys first (most reliable)
        if stats_norm:
            for candidate in [location, display]:
                nk = _norm_for_match(candidate)
                if nk in stats_norm:
                    return stats_norm[nk]
                # Try prefix match: "ohio st" matches "ohio st" in stats
                for sk, sv in stats_norm.items():
                    if sk and nk and (nk.startswith(sk) or sk.startswith(nk)):
                        if abs(len(nk) - len(sk)) <= 4:
                            return sv

        # Fallback: use location (shorter than displayName)
        return location or display

    for groups in (100, 98):
        try:
            url = f"{base}?dates={date_yyyymmdd}&groups={groups}&seasontype=3&limit=200"
            res = requests.get(url, headers={"user-agent": "ncaa-picks-bot/1.0"}, timeout=10)
            if not res.ok:
                continue
            data = res.json()
            for ev in (data.get("events") or []):
                for comp in (ev.get("competitions") or []):
                    for c in (comp.get("competitors") or []):
                        team_obj = c.get("team") or {}
                        display = team_obj.get("displayName")
                        location = team_obj.get("location")
                        if not display:
                            continue
                        best = _best_name(team_obj)
                        # Register multiple normalized variants
                        for variant in [display, location, best]:
                            if variant:
                                teams[_norm_for_match(variant)] = best
        except Exception:
            continue

    return teams


def extract_final_scores(scoreboard_json):
    out = []
    for ev in (scoreboard_json.get("events") or []):
        comp = (ev.get("competitions") or [None])[0]
        if not comp:
            continue

        status = (comp.get("status") or {}).get("type", {}).get("name")
        is_final = status == "STATUS_FINAL" or status == "STATUS_FINAL_OVERTIME"
        if not is_final:
            continue

        competitors = comp.get("competitors") or []
        away = next((c for c in competitors if c.get("homeAway") == "away"), None)
        home = next((c for c in competitors if c.get("homeAway") == "home"), None)
        if not away or not home:
            continue

        away_name = (away.get("team") or {}).get("displayName")
        home_name = (home.get("team") or {}).get("displayName")
        try:
            away_score = float(away.get("score"))
            home_score = float(home.get("score"))
        except (TypeError, ValueError):
            continue
        if not away_name or not home_name or not math.isfinite(away_score) or not math.isfinite(home_score):
            continue

        out.append({
            "away": away_name,
            "home": home_name,
            "awayScore": away_score,
            "homeScore": home_score,
            "date": (scoreboard_json.get("day") or {}).get("date")
        })
    return out

File: jsNCAA/scripts/sources/espn_scoreboard.mjs
// scripts/sources/espn_scoreboard.mjs
// ESPN scoreboard for NCAA men's basketball
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const CACHE_DIR = path.resolve(__dirname, "..", "..", "..", "data", "espn_cache", "ncaab");

export async function fetchScoreboard(dateYYYYMMDD) {
  // Disk cache — shared across NCAA engines (skip for today — statuses change)
  const _today = new Date().toISOString().slice(0, 10).replace(/-/g, "");
  if (dateYYYYMMDD && dateYYYYMMDD !== _today) {
    if (!fs.existsSync(CACHE_DIR)) fs.mkdirSync(CACHE_DIR, { recursive: true });
    const diskPath = path.join(CACHE_DIR, dateYYYYMMDD + ".json");
    if (fs.existsSync(diskPath)) {
      const cached = JSON.parse(fs.readFileSync(diskPath, "utf8"));
      const allFinal = (cached.events || []).length > 0 && (cached.events || []).every(ev => {
        const st = ev?.competitions?.[0]?.status?.type?.name || "";
        return st === "STATUS_FINAL" || st === "STATUS_FINAL_OVERTIME";
      });
      if (allFinal) return cached;
    }
  }

  const base = "https://site.api.espn.com/apis/site/v2/sports/basketball/mens-college-basketball/scoreboard";
  const params = dateYYYYMMDD ? `?dates=${dateYYYYMMDD}&groups=50&limit=365` : `?groups=50&limit=365`;
  const url = `${base}${params}`;
  const res = await fetch(url, { headers: { "user-agent": "ncaa-picks-bot/1.0" } });
  if (!res.ok) throw new Error(`HTTP ${res.status} for ${url}`);
  const data = await res.json();

  if (dateYYYYMMDD) {
    try { fs.writeFileSync(path.join(CACHE_DIR, dateYYYYMMDD + ".json"), JSON.stringify(data)); }
    catch (e) { /* ignore */ }
  }

  return data;
}

export function extractFinalScores(scoreboardJson) {
  const out = [];
  for (const ev of (scoreboardJson.events || [])) {
    const comp = ev.competitions?.[0];
    if (!comp) continue;

    const status = comp.status?.type?.name;
    const isFinal = status === "STATUS_FINAL" || status === "STATUS_FINAL_OVERTIME";
    if (!isFinal) continue;

    const competitors = comp.competitors || [];
    const away = competitors.find(c => c.homeAway === "away");
    const home = competitors.find(c => c.homeAway === "home");
    if (!away || !home) continue;

    const awayName = away.team?.displayName;
    const homeName = home.team?.displayName;
    const awayScore = parseFloat(away.score);
    const homeScore = parseFloat(home.score);
    if (!awayName || !homeName || !Number.isFinite(awayScore) || !Number.isFinite(homeScore)) continue;

    out.push({ away: awayName, home: homeName, awayScore, homeScore, date: scoreboardJson.day?.date || null });
  }
  return out;
}

function _normForMatch(s) {
  return String(s || "").toLowerCase()
    .replace(/[^a-z0-9\s]/g, " ")
    .replace(/\bstate\b/g, "st")
    .replace(/\s+/g, " ").trim();
}

export async function fetchTournamentTeams(dateYYYYMMDD, statsKeys = null) {
  const base = "https://site.api.espn.com/apis/site/v2/sports/basketball/mens-college-basketball/scoreboard";
  const teams = {};

  // Build stats lookup for resolving ESPN names to model names
  const statsNorm = {};
  if (statsKeys) {
    for (const k of statsKeys) {
      statsNorm[_normForMatch(k)] = k;
    }
  }

  function bestName(teamObj) {
    const location = (teamObj.location || "").trim();
    const display = (teamObj.displayName || "").trim();

    if (Object.keys(statsNorm).length) {
      for (const candidate of [location, display]) {
        const nk = _normForMatch(candidate);
        if (statsNorm[nk]) return statsNorm[nk];
        for (const [sk, sv] of Object.entries(statsNorm)) {
          if (sk && nk && (nk.startsWith(sk) || sk.startsWith(nk))) {
            if (Math.abs(nk.length - sk.length) <= 4) return sv;
          }
        }
      }
    }
    return location || display;
  }

  for (const groups of [100, 98]) {
    try {
      const url = `${base}?dates=${dateYYYYMMDD}&groups=${groups}&seasontype=3&limit=200`;
      const res = await fetch(url, { headers: { "user-agent": "ncaa-picks-bot/1.0" } });
      if (!res.ok) continue;
      const data = await res.json();
      for (const ev of (data.events || [])) {
        for (const comp of (ev.competitions || [])) {
          for (const c of (comp.competitors || [])) {
            const teamObj = c.team || {};
            const display = teamObj.displayName;
            const location = teamObj.location;
            if (!display) continue;
            const best = bestName(teamObj);
            for (const variant of [display, location, best]) {
              if (variant) teams[_normForMatch(variant)] = best;
            }
          }
        }
      }
    } catch (e) { continue; }
  }
  return teams;
}
